### PR TITLE
Move to platform background work indicator

### LINKED ISSUE #63619: Symbol renaming takes forever due to the wrong MEF import of IToolTipPresenterFactory
**Version Used**: 

Visual Studio 17.3.2

**Steps to Reproduce**:

1. Create a VS extension with the following MEF export:
   ```csharp
   [Export(typeof(IToolTipPresenterFactory))]
   sealed class ToolTipPresenterFactory : IToolTipPresenterFactory
   {
       public IToolTipPresenter Create(ITextView textView, ToolTipParameters parameters) => null;
   }
   ```
2. Run a VS instance with the extension installed.
3. Create a new CSharp project and rename any symbol: method, variable, etc.

**Expected Behavior**:

Renaming succeeds immediately.

**Actual Behavior**:

A `Computing Rename information...` dialog appears and VS completely hangs.

---

The constructor of [`Microsoft.CodeAnalysis.Editor.BackgroundWorkIndicator.WpfBackgroundWorkIndicatorFactory`](https://github.com/dotnet/roslyn/blob/1ce74d634ec9cfb24003fa01f04877a35accd79f/src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/WpfBackgroundWorkIndicatorFactory.cs) class imports [`Microsoft.VisualStudio.Text.Adornments.IToolTipPresenterFactory`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.text.adornments.itooltippresenterfactory?view=visualstudiosdk-2022) interface like it always has a single implementation. Exporting a second implementation leads to the issue. A VS internal `Microsoft.VisualStudio.Text.AdornmentLibrary.ToolTip.Implementation.ToolTipService` uses `[ImportMany]` as expected.

There was no such issue before VS 17.3.

### PR DESCRIPTION
**Pull request by ryzngard**
Fixes #63619 

This still keeps us wrapping inside Roslyn, in cases where we won't have a WPF export.

### CODE REVIEW

File: src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/WpfBackgroundWorkIndicatorFactory.cs [hunk 1]
```diff
@@ -48,22 +45,32 @@ internal sealed partial class WpfBackgroundWorkIndicatorFactory : IBackgroundWor
             _threadingContext.ThrowIfNotOnUIThread();
 
             // If we have an outstanding context in flight, cancel it and create a new one to show the user.
-            _currentContext?.CancelAndDispose();
+            _currentContext?.Dispose();
```
Comment (ryzngard) [inline src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/WpfBackgroundWorkIndicatorFactory.cs:48]: The platform supports multiple, so we can make the decision if we want multiple to be shown

File: src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/BackgroundWorkIndicatorContext.cs [hunk 2]
```diff
@@ -21,281 +21,54 @@ namespace Microsoft.CodeAnalysis.Editor.BackgroundWorkIndicator;
...
-        /// </summary>
-        public void TakeOwnership()
-            => this.Dispose();
-
-        private void OnTextBufferChanged(object? sender, TextContentChangedEventArgs e)
-        {
-            if (CancelOnEdit)
-                CancelAndDispose();
-        }
+        public PropertyCollection Properties { get; } = new PropertyCollection();
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/BackgroundWorkIndicatorContext.cs:37]: i'd be ok throwing on this.  Is anyone actually using this?

File: src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/BackgroundWorkIndicatorScope.cs [hunk 3]
```diff
@@ -3,88 +3,49 @@
...
-        }
-
-        /// <summary>
-        /// On disposal, just remove ourselves from our parent context.  It will update the UI accordingly.
-        /// </summary>
-        void IDisposable.Dispose()
-            => _context.RemoveScope(this);
-
-        bool IUIThreadOperationScope.AllowCancellation
+        private class BackgroundWorkIndicatorScope : IUIThreadOperationScope
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/BackgroundWorkIndicatorScope.cs:12]: do we need our own impl here?  or is this because the provided wait-indicator has no way to create scopes?

File: src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/BackgroundWorkIndicatorContext.cs [hunk 4]
```diff
@@ -21,281 +21,54 @@ namespace Microsoft.CodeAnalysis.Editor.BackgroundWorkIndicator;
...
-                Contract.ThrowIfFalse(_scopes.Contains(scope));
-                _scopes = _scopes.Remove(scope);
-            }
-
-            // We changed.  Enqueue work to make sure the UI reflects this.
-            this.EnqueueUIUpdate();
-        }
-
-        private (string description, ProgressInfo progressInfo) BuildData()
+        public void TakeOwnership()
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/BackgroundWorkIndicatorContext.cs:64]: do we use this?

File: src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/BackgroundWorkIndicatorScope.cs [hunk 5]
```diff
@@ -3,88 +3,49 @@
...
             {
-                lock (_context.Gate)
-                    return _description;
+                _backgroundScope = backgroundScope;
+                _description = description;
+                _backgroundWorkIndicatorContext = backgroundWorkIndicatorContext;
             }
-            set
+
+            public bool AllowCancellation { get; set; }
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/BackgroundWorkIndicatorScope.cs:27]: all these props seem odd.  i would expect we would be forwarding to the underlying BackgroundWorkOperationScope  we're wrapping.

File: src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/BackgroundWorkIndicatorContext.cs [hunk 6]
```diff
@@ -21,281 +21,54 @@ namespace Microsoft.CodeAnalysis.Editor.BackgroundWorkIndicator;
...
-                CancelAndDispose();
-        }
+        public PropertyCollection Properties { get; } = new PropertyCollection();
 
-        private void OnTextViewLostAggregateFocus(object? sender, EventArgs e)
-        {
-            if (CancelOnFocusLost)
-                CancelAndDispose();
-        }
+        public event EventHandler? OnDisposed;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/BackgroundWorkIndicatorContext.cs:39]: i would prefer to not expose this generic event.  This is only made by the factory, so the factory can pass itself into this.  When the context is disposed, it can tell the factory directly.

Comment (ryzngard) [reply src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/BackgroundWorkIndicatorContext.cs:39]: Works for me. I wasn't sure we wanted that strong of a tie in but I have no qualms with it.

File: src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/BackgroundWorkIndicatorScope.cs (hunk 3 above)
Comment (ryzngard) [reply src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/BackgroundWorkIndicatorScope.cs:12]: The implementation from platform doesn't inherit from `IUIThreadOperationScope`, just `IDisposable`. This gets into the "Do we want an abstraction"  question. It would require more changes to get rid of that abstraction, but may be worth it.

File: src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/BackgroundWorkIndicatorContext.cs (hunk 4 above)
Comment (ryzngard) [reply src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/BackgroundWorkIndicatorContext.cs:64]: Vestigial from us merging with `IUIThreadOperationScope` even though the platform doesn't. This is a rough edge we have to deal with if we want complete abstraction. I'm happy to not have out own wrappers and move to just the platform implementation

File: src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/BackgroundWorkIndicatorScope.cs (hunk 5 above)
Comment (ryzngard) [reply src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/BackgroundWorkIndicatorScope.cs:27]: All part of `IUIThreadOperationScope` :(

File: src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/BackgroundWorkIndicatorScope.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/BackgroundWorkIndicatorScope.cs:12]: That's... interesting.  How do teh scopes work?  @kayle can you point to the docs on that?

Comment (kayle) [reply src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/BackgroundWorkIndicatorScope.cs:12]: They're very basic right now. You can create/dispose them and update their descriptions. The current presentation will show the description of the last (not-disposed) scope:
```
    /// <summary>
    /// Represents a single scope in the context of executing a potentially long running operation.
    /// </summary>
    /// <remarks>Members can be invoked from any thread.</remarks>
    public abstract class BackgroundWorkOperationScope : IDisposable
    {
        internal BackgroundWorkOperationScope() { } // Prevent external implementations

        /// <summary>
        /// Updates the description of the operation scope.
        /// </summary>
        public abstract string Description { set; }

        /// <summary>
        /// Marks the operation scope as completed
        /// </summary>
        public abstract void Dispose();
    }
```

I do expect we'll add progress reporting in the future, but possibly not for operations owned by a TextView. We will likely not mix the IUIThreadOperationScope with the BackgroundWorkOperationScope or any future entry points on IBackgroundWorkIndicatorService.

File: src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/WpfBackgroundWorkIndicatorFactory.cs [hunk 7]
```diff
@@ -21,21 +21,18 @@ namespace Microsoft.CodeAnalysis.Editor.BackgroundWorkIndicator
     internal sealed partial class WpfBackgroundWorkIndicatorFactory : IBackgroundWorkIndicatorFactory
```
Comment (kayle) [inline src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/WpfBackgroundWorkIndicatorFactory.cs:21]: The platform implementation should be cross platform. Could you make this the default implementation now and delete your other context?

File: src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/BackgroundWorkIndicatorScope.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/BackgroundWorkIndicatorScope.cs:12]: >We will likely not mix the IUIThreadOperationScope with the BackgroundWorkOperationScope

I do find that very very odd.  From my perspective, these are just a different, lightweight, view on the same model and controller.  Having a different API with disparate functionality seems odd (esp. as the functionality seems relevant to all views).

Comment (ryzngard) [reply src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/BackgroundWorkIndicatorScope.cs:12]: When we discussed this previously, we acknowledged that not everything in `IUIThreadOperationScope` made sense to move forward with. I believe it was `AllowCancellation` specifically that didn't make sense for BWI

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/BackgroundWorkIndicatorScope.cs:12]: >  I believe it was AllowCancellation specifically that didn't make sense for BWI

Hrmm... that seems odd to me.  We have operations (like applying a solution-change) that need to switch over to being non-cancellable once they start actually making the change.  This way the user doesn't end up in some sort of partial-application state.

Comment (ryzngard) [reply src/EditorFeatures/Core.Wpf/BackgroundWorkIndicator/BackgroundWorkIndicatorScope.cs:12]: There is a way to disallow cancellation, I think it was just meant to not be a property. I would have to digup e-mail for exact reasoning

**Comment by CyrusNajmabadi** [comment]
Can you remind me on monday?

**Comment by ryzngard** [comment]
Ping @CyrusNajmabadi

File: src/EditorFeatures/Core.Wpf/NavigableSymbols/NavigableSymbolService.NavigableSymbol.cs [hunk 8]
```diff
@@ -55,13 +52,16 @@ private async Task NavigateAsync()
             {
                 // we're about to navigate.  so disable cancellation on focus-lost in our indicator so we don't end up
                 // causing ourselves to self-cancel.
-                using var backgroundIndicator = _indicatorFactory.Create(
+                using var backgroundIndicator = _service._backgroundWorkIndicatorService.Create(
                     _textView, SymbolSpan,
                     EditorFeaturesResources.Navigating_to_definition,
-                    cancelOnFocusLost: false);
+                    new()
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/NavigableSymbols/NavigableSymbolService.NavigableSymbol.cs:58]: nit: woudl prefer a named type here.  but up to you.

File: src/EditorFeatures/Core.Wpf/NavigableSymbols/NavigableSymbolService.NavigableSymbol.cs [hunk 9]
```diff
@@ -55,13 +52,16 @@ private async Task NavigateAsync()
...
+                    {
+                        CancelOnFocusLost = false
+                    });
 
                 await _location.TryNavigateToAsync(
-                    _service._threadingContext, new NavigationOptions(PreferProvisionalTab: true, ActivateTab: true), backgroundIndicator.UserCancellationToken).ConfigureAwait(false);
+                    _service._threadingContext, new NavigationOptions(PreferProvisionalTab: true, ActivateTab: true), backgroundIndicator.CancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/NavigableSymbols/NavigableSymbolService.NavigableSymbol.cs:64]: @kayle curious why the change in naming here.  seems inconsistent with the TWD api.

File: src/EditorFeatures/Core/EncapsulateField/AbstractEncapsulateFieldCommandHandler.cs [hunk 10]
```diff
@@ -95,11 +100,6 @@ public bool ExecuteCommand(EncapsulateFieldCommandArgs args, CommandExecutionCon
             {
                 await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
 
-                // We are about to show a modal UI dialog so we should take over the command execution
-                // wait context. That means the command system won't attempt to show its own wait dialog 
-                // and also will take it into consideration when measuring command handling duration.
-                context.TakeOwnership();
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/EncapsulateField/AbstractEncapsulateFieldCommandHandler.cs:101]: interesting.  what happens here?  shoul.d we explicitly dispose the context instead?

File: src/EditorFeatures/Core/ExtractMethod/ExtractMethodCommandHandler.cs [hunk 11]
```diff
@@ -210,14 +216,14 @@ public bool ExecuteCommand(ExtractMethodCommandArgs args, CommandExecutionContex
...
-            ITextBuffer textBuffer, IEnumerable<TextChange> changes, IBackgroundWorkIndicatorContext waitContext)
+            ITextBuffer textBuffer, IEnumerable<TextChange> changes, IBackgroundWorkIndicator workIndicator)
         {
             _threadingContext.ThrowIfNotOnUIThread();
 
             using var undoTransaction = _undoManager.GetTextBufferUndoManager(textBuffer).TextBufferUndoHistory.CreateTransaction("Extract Method");
 
             // We're about to make an edit ourselves.  so disable the cancellation that happens on editing.
-            waitContext.CancelOnEdit = false;
+            using var _ = workIndicator.SuppressAutoCancel();
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/ExtractMethod/ExtractMethodCommandHandler.cs:226]: nice.  i like thsi pattern.

File: src/EditorFeatures/Core/Shared/Extensions/ITextSnapshotExtensions.cs [hunk 12]
```diff
@@ -88,6 +88,40 @@ internal static partial class ITextSnapshotExtensions
...
+                {
+                    // TODO: decide for prototype, we don't do anything complex and just ask workspace whether it is fully loaded
+                    // later we might need to go and change all these with more specific info such as document/project/solution
+                    await service.WaitUntilFullyLoadedAsync(backgroundWorkIndicator.CancellationToken).ConfigureAwait(false);
+                }
+
+                // get proper document
+                return snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            }
+        }
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Shared/Extensions/ITextSnapshotExtensions.cs:123]: feels like we can share more code here.  like using that pattern you had of passing around a delegate which calls .AddScope.  LMK if htat works.

**Review by CyrusNajmabadi — APPROVED** [review]
approved, but if you can, please address suggestions.

File: src/EditorFeatures/Core.Wpf/NavigableSymbols/NavigableSymbolService.cs [hunk 13]
```diff
@@ -25,16 +25,19 @@ internal sealed partial class NavigableSymbolService : INavigableSymbolSourcePro
         private readonly IUIThreadOperationExecutor _uiThreadOperationExecutor;
         private readonly IThreadingContext _threadingContext;
         private readonly IAsynchronousOperationListener _listener;
+        private readonly IBackgroundWorkIndicatorService _backgroundWorkIndicatorService;
 
         [ImportingConstructor]
         [SuppressMessage("RoslynDiagnosticsReliability", "RS0033:Importing constructor should be [Obsolete]", Justification = "Used in test code: https://github.com/dotnet/roslyn/issues/42814")]
```
Comment (sharwell) [inline src/EditorFeatures/Core.Wpf/NavigableSymbols/NavigableSymbolService.cs:31]: 📝 This seems like a good time to fix this

File: src/EditorFeatures/Core/EncapsulateField/AbstractEncapsulateFieldCommandHandler.cs (hunk 10 above)
Comment (ryzngard) [reply src/EditorFeatures/Core/EncapsulateField/AbstractEncapsulateFieldCommandHandler.cs:101]: We could add an extension for TakeOwnership that does a dispose... but does that present problems for double dispose?

File: src/EditorFeatures/Core/ExtractMethod/ExtractMethodCommandHandler.cs (hunk 11 above)
Comment (sharwell) [reply src/EditorFeatures/Core/ExtractMethod/ExtractMethodCommandHandler.cs:226]: Are we sure this method applies to cancel on edit? There is also a feature to automatically cancel operations which take too much time and it's not clear from the method name which one is intended.

File: src/EditorFeatures/Core/GoToDefinition/GoToDefinitionCommandHandler.cs [hunk 14]
```diff
@@ -161,32 +163,30 @@ private static void ReportFailure(Document document)
...
-                var cancellationToken = backgroundIndicator.UserCancellationToken;
+                var cancellationToken = backgroundIndicator.CancellationToken;
 
                 // determine the location first.
                 var (location, _) = await service.FindDefinitionLocationAsync(
                     document, position, includeType: true, cancellationToken).ConfigureAwait(false);
 
                 // make sure that if our background indicator got canceled, that we do not still perform the navigation.
-                if (backgroundIndicator.UserCancellationToken.IsCancellationRequested)
+                if (backgroundIndicator.CancellationToken.IsCancellationRequested)
```
Comment (sharwell) [inline src/EditorFeatures/Core/GoToDefinition/GoToDefinitionCommandHandler.cs:184]: Can this be simplified?

```suggestion
                if (cancellationToken)
```

File: src/EditorFeatures/Core/GoToDefinition/GoToDefinitionCommandHandler.cs [hunk 15]
```diff
@@ -161,32 +163,30 @@ private static void ReportFailure(Document document)
...
                     return;
 
                 // we're about to navigate.  so disable cancellation on focus-lost in our indicator so we don't end up
                 // causing ourselves to self-cancel.
-                backgroundIndicator.CancelOnFocusLost = false;
```
Comment (sharwell) [inline src/EditorFeatures/Core/GoToDefinition/GoToDefinitionCommandHandler.cs:189]: Does `CancelOnFocusLost` default to true or false? I see cases above that explicitly pass `CancelOnFocusLost = true`, yet here we did not explicitly state that and are now setting `CancelOnFocusLost = false`. It seems like one of the two is redundant.

File: src/EditorFeatures/Core/InlineRename/CommandHandlers/AbstractRenameCommandHandler_RenameHandler.cs [hunk 16]
```diff
@@ -22,6 +21,13 @@ namespace Microsoft.CodeAnalysis.Editor.Implementation.InlineRename
 {
     internal abstract partial class AbstractRenameCommandHandler : ICommandHandler<RenameCommandArgs>
     {
+        private readonly IBackgroundWorkIndicatorService _backgroundWorkIndicatorService;
+
+        internal AbstractRenameCommandHandler(IBackgroundWorkIndicatorService backgroundWorkIndicatorService)
```
Comment (sharwell) [inline src/EditorFeatures/Core/InlineRename/CommandHandlers/AbstractRenameCommandHandler_RenameHandler.cs:26]:
```suggestion
        protected AbstractRenameCommandHandler(IBackgroundWorkIndicatorService backgroundWorkIndicatorService)
```

File: src/EditorFeatures/TestUtilities/Utilities/TestBackgroundWorkIndicatorService.cs [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Utilities;
+using Moq;
+
+namespace Microsoft.CodeAnalysis.Test.Utilities.Utilities
+{
+    internal class TestBackgroundWorkIndicatorService : IBackgroundWorkIndicatorService
```
Comment (sharwell) [inline src/EditorFeatures/TestUtilities/Utilities/TestBackgroundWorkIndicatorService.cs:18]: ❗ This should not be directly constructed in tests. It should be placed in the `TestComposition` and accessed through MEF.

```suggestion
    [Export(typeof(IBackgroundWorkIndicatorService))]
    [Shared]
    [PartNotDiscoverable]
    internal class TestBackgroundWorkIndicatorService : IBackgroundWorkIndicatorService
```

File: src/EditorFeatures/Core/ExtractMethod/ExtractMethodCommandHandler.cs (hunk 11 above)
Comment (ryzngard) [reply src/EditorFeatures/Core/ExtractMethod/ExtractMethodCommandHandler.cs:226]: This applies to the BackgroundWorkIndicatorOptions set on the indicator

File: src/EditorFeatures/Core/GoToDefinition/GoToDefinitionCommandHandler.cs (hunk 15 above)
Comment (ryzngard) [reply src/EditorFeatures/Core/GoToDefinition/GoToDefinitionCommandHandler.cs:189]: I'd have to check implementation, but I'm not opposed to explicitly setting it on every creation. My guess it defaults to false because they are `{ get; init; }` defined. @kayle ?

**Comment by allisonchou** [comment]
@ryzngard Are we able to safely re-target this PR to main? We're attempting to delete the -vs-deps branches soon.

**Comment by CyrusNajmabadi** [comment]
Please don't delete -vs-deps.  I will need it this sprint as part of the AIOSP work.  Thanks!
